core/runtime/tests/jsrealm.rs

```diff
@@ -803,3 +803,392 @@ async fn test_realm_concurrent_dynamic_imports() {
+            "export const TEST = 'foo';",
+          ),
+        },
+        ExtensionFileSource {
+          specifier: "test:not-preserved",
+          code: ExtensionFileSourceCode::IncludedInBinary(
+            "import 'test:preserved'; export const TEST = 'bar';",
+          ),
+        },
+      ])
+      .esm_entry_point("test:not-preserved")
+      .build();
+
+    let runtime = JsRuntimeForSnapshot::new(
+      RuntimeOptions {
+        extensions: vec![extension],
+        ..Default::default()
+      },
+      Default::default(),
+    );
+
+    runtime.snapshot()
+  };
+
+  struct Loader;
+
+  impl ModuleLoader for Loader {
+    fn resolve(
+      &self,
+      specifier: &str,
+      referrer: &str,
+      _kind: ResolutionKind,
+    ) -> Result<ModuleSpecifier, Error> {
+      Ok(resolve_import(specifier, referrer)?)
+    }
+
+    fn load(
+      &self,
+      module_specifier: &ModuleSpecifier,
+      _maybe_referrer: Option<&ModuleSpecifier>,
+      _is_dyn_import: bool,
+    ) -> Pin<Box<ModuleSourceFuture>> {
+      assert_eq!(module_specifier.as_str(), "test:not-preserved");
+      futures::future::ready(Err(error::generic_error("Couldn't load module")))
+        .boxed_local()
+    }
+  }
+
+  let loader = Rc::new(Loader);
+
+  let mut runtime = JsRuntime::new(RuntimeOptions {
+    module_loader: Some(loader.clone()),
+    startup_snapshot: Some(Snapshot::JustCreated(startup_data)),
+    preserve_snapshotted_modules: Some(&["test:preserved"]),
+    ..Default::default()
+  });
+
+  let realm = runtime
+    .create_realm(CreateRealmOptions {
+      module_loader: Some(loader),
+    })
+    .unwrap();
+
+  // We can't import "test:not-preserved"
+  {
+    let dyn_import_promise = realm
+      .execute_script_static(
+        runtime.v8_isolate(),
+        "",
+        "import('test:not-preserved')",
+      )
+      .unwrap();
+    let dyn_import_result =
+      futures::executor::block_on(runtime.resolve_value(dyn_import_promise));
+    assert!(dyn_import_result.is_err());
+    assert_eq!(
+      dyn_import_result.err().unwrap().to_string().as_str(),
+      "Uncaught TypeError: Couldn't load module"
+    );
+  }
+
+  // But we can import "test:preserved"
+  {
+    let dyn_import_promise = realm
+      .execute_script_static(
+        runtime.v8_isolate(),
+        "",
+        "import('test:preserved').then(module => module.TEST)",
+      )
+      .unwrap();
+    let dyn_import_result =
+      futures::executor::block_on(runtime.resolve_value(dyn_import_promise))
+        .unwrap();
+    let scope = &mut realm.handle_scope(runtime.v8_isolate());
+    assert!(dyn_import_result.open(scope).is_string());
+    assert_eq!(
+      dyn_import_result
+        .open(scope)
+        .to_rust_string_lossy(scope)
+        .as_str(),
+      "foo"
+    );
+  }
+}
+
+/// Test that `RuntimeOptions::preserve_snapshotted_modules` also works without
+/// a snapshot.
+#[test]
+fn non_snapshot_preserve_snapshotted_modules() {
```

Comment:
For tests that are identical other than a small change, let's try to avoid copy+paste. Given that this particular test is already kind of combinatorial, maybe we can have a base test that takes two parameters: snapshot: bool and realm: bool.